File: api/app/model/post.go
package model

import "time"

type Post struct {
	ID        int       `json:"id"`
	UserID    int       `json:"user_id"`
	Title     string    `json:"title"`
	Img       *string   `json:"img"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type IndexPost struct {
	Post Post
	User User
}

type ShowPost struct {
	IndexPost         IndexPost
	CommenstWithUsers []*ShowCommentWithUser
	Likes             []*Like
}

type IndexPostWithCountLike struct {
	IndexPost    IndexPost
	CountLike    int
	CountComment int
}

type ShowCommentWithUser struct {
	Comment CommentNil
	User    UserNil
}

File: client/lib/model/post.ts
import { CommentWithUser } from "./comment";
import { Like } from "./like";
import { User } from "./user";
export class Post {
  constructor(
    public id: number,
    public userId: number,
    public title: string,
    public createdAt: Date,
    public updatedAt: Date,
    public img?: string,
  ) {}
}

export interface SelectPost {
  id: number;
  title: string;
  img?: string;
}

export interface PostWithUser {
  post: Post;
  user: Omit<User, "id" | "email" | "email" | "password" | "createdAt">;
  countLike: number;
  countComment: number;
}

export interface ShowPost {
  post: Post;
  user: Omit<User, "email" | "password" | "createdAt">;
  likes: Like[];
  commentsWithUsers: CommentWithUser[];
}

export interface PostAllLimit {
  postsWithUsers: PostWithUser[];
  nextId: number | null;
}
